Mathlib/Data/Set/Finite/Basic.lean

```diff
@@ -688,6 +688,22 @@ theorem Finite.dinduction_on {C : ∀ s : Set α, s.Finite → Prop} (s : Set α
     Finite.induction_on h (fun _ => H0) fun has hs ih _ => H1 has hs (ih _)
   this h
 
+theorem setOf_nodup_perm (l : List α) (H : l.Nodup) [DecidableEq α] :
+    l.finite_toSet.toFinset.toList.Perm l := by
+  rw [List.perm_iff_count]
+  intro a
+  by_cases h : a ∈ l
+  · rw [l.count_eq_one_of_mem H h,
+      l.finite_toSet.toFinset.toList.count_eq_one_of_mem (Finset.nodup_toList _) (by simp_all)]
+  · rw [l.count_eq_zero_of_not_mem h,
+      l.finite_toSet.toFinset.toList.count_eq_zero_of_not_mem (by simp_all)]
+
+theorem setOf_mem_list_eq_singleton_of_nodup {l : List α} (H : l.Nodup) {a : α} [DecidableEq α] :
```

Comment:
You can state this result without `DecidableEq α` and instead start the proof with `classical`, which should hopefully make the weird `Fintype {x | x ∈ l}` instance unnecessary.

Reply:
I opted not to do this, as inferring the non-simp-normal `Finite` instance was still necessary even with classical.

Reply:
Yes `DecidableEq` should be removed from both the statements here as it is not needed to make the statement typecheck. As Violeta says, the proofs should start `classical`.